File: tools/build-contract.mjs
#!/usr/bin/env node
/**
 * Livello 2 — contratto di componente, estratto da Figma.
 *
 * Sostituisce build-contract-from-metadata.mjs, che ricavava gli assi dalla
 * PROSA di `component.description` (e infatti sbagliava default e ordine).
 * Qui la sorgente è la property table reale più la geometria dei nodi.
 *
 * Due principi:
 *  · OUTPUT CANONICO — chiavi ordinate, float arrotondati, nessun timestamp
 *    dentro l'artefatto: ogni diff è un cambiamento vero, mai rumore.
 *  · NON PROMETTE PIÙ DI QUANTO HA MISURATO — ogni sezione dichiara su quante
 *    varianti è stata verificata.
 *
 *   node build-contract.mjs <slug> <nodeId>
 *   node build-contract.mjs button 5473-10855
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT, TOKEN } from './_env.mjs';

const OUT = `${ROOT}/data/contracts`;

/* Il file NON è più scritto qui dentro.
   I due DS condividono gli id dei nodi (il B2B è un duplicato del B2C), quindi
   lo stesso nodeId su file diversi restituisce componenti diversi: il ds deve
   essere esplicito, altrimenti si documenta il componente sbagliato. */
const CFG = JSON.parse(fs.readFileSync(`${ROOT}/config/figma-files.json`, 'utf8'));

const slug = process.argv[2] ?? 'button';
const nodeId = process.argv[3] ?? '5473-10855';
const ds = process.argv[4] ?? CFG.default;

if (!CFG.files[ds]) {
  console.error(`\n  ✖ ds sconosciuto: "${ds}" — attesi: ${Object.keys(CFG.files).join(', ')}\n`);
  process.exit(1);
}
const FILE = CFG.files[ds].fileKey;

const api = async u => {
  const r = await fetch(`https://api.figma.com/v1/${u}`, { headers: { 'X-Figma-Token': TOKEN } });
  if (!r.ok) throw new Error(`Figma HTTP ${r.status} — ${u}`);
  return r.json();
};

/* ── mappa variableId → nome, per tradurre i boundVariables ──────────────── */
const { meta: vmeta } = await api(`files/${FILE}/variables/local`);
const varName = id => vmeta.variables[id]?.name ?? null;

/* ── il component set ─────────────────────────────────────────────────────── */
const nodes = await api(`files/${FILE}/nodes?ids=${nodeId}`);
const doc = Object.values(nodes.nodes)[0].document;
const variants = doc.children ?? [];

/* ── 1. Assi: dalla property table, non dalla prosa ──────────────────────── */
const defs = doc.componentPropertyDefinitions ?? {};
const clean = n => n.split('#')[0].trim();          // "Show left icon#5615:73" → "Show left icon"
const camel = n => { const p = clean(n).split(/[\s_-]+/); return p[0].toLowerCase() + p.slice(1).map(w => w[0].toUpperCase() + w.slice(1)).join(''); };

const axes = [], booleans = [], texts = [], swaps = [];
for (const [rawName, d] of Object.entries(defs)) {
  const entry = { figmaProperty: clean(rawName), name: camel(rawName) };
  if (d.type === 'VARIANT') axes.push({ ...entry, values: [...d.variantOptions].sort() });
  else if (d.type === 'BOOLEAN') booleans.push({ ...entry, default: d.defaultValue });
  else if (d.type === 'TEXT') texts.push({ ...entry, default: d.defaultValue });
  else swaps.push({ ...entry, type: d.type });
}

/* ── 2. Il default: la PRIMA variante del set è il riferimento di Figma ──── */
const parseVariant = name => Object.fromEntries(
  name.split(',').map(s => s.split('=').map(x => x.trim())).filter(p => p.length === 2));
const firstCombo = variants.length ? parseVariant(variants[0].name) : {};

/* ── 3. Sizing e token: dai boundVariables dei nodi variante ─────────────── */
const round = n => typeof n === 'number' ? Math.round(n * 100) / 100 : n;

function surfacesOf(node) {
  const out = {};
  const bv = node.boundVariables ?? {};
  const fill = node.fills?.[0]?.boundVariables?.color?.id;
  const stroke = node.strokes?.[0]?.boundVariables?.color?.id;
  if (fill) out.bg = varName(fill);
  if (stroke) out.border = varName(stroke);
  // testo e icona stanno nei figli
  const walk = n => {
    for (const c of n.children ?? []) {
      const cf = c.fills?.[0]?.boundVariables?.color?.id;
      if (cf) { if (c.type === 'TEXT' && !out.text) out.text = varName(cf); else if (!out.icon) out.icon = varName(cf); }
      walk(c);
    }
  };
  walk(node);
  if (bv.strokeWeight) out._borderWidth = varName(bv.strokeWeight.id);
  return out;
}

function geometryOf(node) {
  const bv = node.boundVariables ?? {};
  const g = {};
  const map = { paddingTop: 'paddingVertical', paddingLeft: 'paddingHorizontal',
                itemSpacing: 'gap', strokeWeight: 'borderWidth' };
  for (const [k, label] of Object.entries(map)) if (bv[k]) g[label] = varName(bv[k].id);
  if (bv.bottomLeftRadius || bv.topLeftRadius)
    g.borderRadius = varName((bv.topLeftRadius ?? bv.bottomLeftRadius).id);
  g._measured = { height: round(node.absoluteBoundingBox?.height), layout: node.layoutMode ?? null };
  return g;
}

const sizeAxis = axes.find(a => /size/i.test(a.name));
const appearAxis = axes.find(a => /appear/i.test(a.name));
const hierAxis = axes.find(a => /hierarch/i.test(a.name));
const stateAxis = axes.find(a => /state/i.test(a.name));

const sizing = {}, tokens = {};
const coverage = { variantsTotal: variants.length, sizingFrom: 0, tokensFrom: 0 };

for (const v of variants) {
  const combo = parseVariant(v.name);
  // sizing: una volta per size, dalla combinazione di default degli altri assi
  const sz = sizeAxis && combo[sizeAxis.figmaProperty];
  if (sz && !sizing[sz]) { sizing[sz] = geometryOf(v); coverage.sizingFrom++; }

  // tokens: per ogni (appearance, hierarchy) × state
  if (appearAxis && hierAxis && stateAxis) {
    const key = `${combo[appearAxis.figmaProperty]}/${combo[hierAxis.figmaProperty]}`;
    const st = combo[stateAxis.figmaProperty];
    if (key.includes('undefined') || !st) continue;
    // solo dalla size di riferimento, per non moltiplicare
    if (sizeAxis && combo[sizeAxis.figmaProperty] !== firstCombo[sizeAxis.figmaProperty]) continue;
    (tokens[key] ??= {})[st] = surfacesOf(v);
    coverage.tokensFrom++;
  }
}

/* ── 4. Il contratto, in forma canonica ──────────────────────────────────── */
const sortKeys = o => Array.isArray(o) ? o.map(sortKeys)
  : (o && typeof o === 'object')
    ? Object.fromEntries(Object.keys(o).sort().map(k => [k, sortKeys(o[k])]))
    : o;

const contract = {
  id: `ds.${slug}`,
  name: doc.name,
  source: { ds, fileKey: FILE, fileName: CFG.files[ds].name, nodeId, nodeType: doc.type },

  props: axes.map(a => ({
    name: a.name,
    figmaProperty: a.figmaProperty,
    values: a.values,                                   // ordine: alfabetico, come Figma
    default: firstCombo[a.figmaProperty] ?? null,       // la prima variante del set
    _source: { values: 'figma', order: 'figma-alphabetical', default: 'figma-first-variant' },
  })),
  booleans, texts, swaps,

  sizing,
  tokens,

  _coverage: {
    variants: coverage.variantsTotal,
    sizingMeasuredOn: `${coverage.sizingFrom}/${sizeAxis?.values.length ?? 0} size`,
    tokensMeasuredOn: `${coverage.tokensFrom} combinazioni su ${(appearAxis?.values.length ?? 0) * (hierAxis?.values.length ?? 0) * (stateAxis?.values.length ?? 0)}`,
    note: 'I token sono misurati sulla sola size di riferimento: le superfici non variano con la size.',
  },
};

fs.mkdirSync(OUT, { recursive: true });
const dir = path.join(OUT, slug);
fs.mkdirSync(dir, { recursive: true });
const file = path.join(dir, 'contract.json');
fs.writeFileSync(file, JSON.stringify(sortKeys(contract), null, 2) + '\n');

/* ── Report ───────────────────────────────────────────────────────────────── */
const L = '─'.repeat(70);
console.log(`\n${L}\n  CONTRATTO da Figma — ${doc.name}\n${L}`);
console.log(`\n▸ NODO      ${doc.type} · ${variants.length} varianti`);
console.log(`\n▸ ASSI      (dalla property table, non dalla prosa)`);
for (const a of axes) console.log(`   ${a.name.padEnd(12)} figma="${a.figmaProperty}"  default=${firstCombo[a.figmaProperty] ?? '—'}\n      ${a.values.join(' | ')}`);
if (booleans.length) console.log(`\n▸ BOOLEAN   ${booleans.map(b => `${b.name}=${b.default}`).join(', ')}`);
if (texts.length) console.log(`▸ TEXT      ${texts.map(t => `${t.name}="${t.default}"`).join(', ')}`);
if (swaps.length) console.log(`▸ SWAP      ${swaps.map(s => s.name).join(', ')}`);
console.log(`\n▸ SIZING    ${Object.keys(sizing).join(', ')}`);
for (const [k, g] of Object.entries(sizing))
  console.log(`   ${k}: ${Object.entries(g).filter(([x]) => !x.startsWith('_')).map(([x, t]) => `${x}=${t}`).join(' · ') || '(nessun binding)'}  [h=${g._measured.height}]`);
console.log(`\n▸ TOKEN     ${Object.keys(tokens).length} combinazioni appearance/hierarchy`);
const sample = Object.entries(tokens)[0];
if (sample) {
  console.log(`   esempio "${sample[0]}":`);
  for (const [st, surf] of Object.entries(sample[1]))
    console.log(`      ${st.padEnd(9)} ${Object.entries(surf).filter(([k]) => !k.startsWith('_')).map(([k, v]) => `${k}=${v}`).join(' · ') || '—'}`);
}
console.log(`\n▸ COPERTURA ${JSON.stringify(contract._coverage, null, 0)}`);
console.log(`\n✅ scritto ${path.relative(ROOT, file)}`);
console.log(`${L}\n`);

File: tools/build_contract.py
#!/usr/bin/env python3
"""
Livello 2 — contratto di componente, estratto da Figma.

La sorgente e' la property table reale piu' la geometria dei nodi, non la PROSA
della descrizione: ricavare gli assi dal testo sbagliava default e ordine.

Due principi:
 · OUTPUT CANONICO — chiavi ordinate, float arrotondati, nessun timestamp
   dentro l'artefatto: ogni diff e' un cambiamento vero, mai rumore.
 · NON PROMETTE PIU' DI QUANTO HA MISURATO — ogni sezione dichiara su quante
   varianti e' stata verificata.

Il file Figma NON e' scritto qui dentro: i due DS condividono gli id dei nodi,
quindi lo stesso nodeId su file diversi restituisce componenti diversi. Il ds
va indicato, e un controllo verifica che il nodo sia quello atteso.

  python tools/build_contract.py <slug> <nodeId> [ds] [--forza]
  python tools/build_contract.py button 5473-10855 b2c
"""
import json
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
from _figma import ROOT, Figma, scrivi_json  # noqa: E402
from _guardia import verifica_nodo  # noqa: E402

USCITA = ROOT / "data" / "contracts"


def arrotonda(n):
    """
    Come `Math.round(n*100)/100` in JavaScript, che scrive 40 e non 40.0.
    Serve perche' gli artefatti gia' prodotti vanno riprodotti identici.
    """
    if not isinstance(n, (int, float)) or isinstance(n, bool):
        return n
    v = round(float(n) * 100) / 100
    return int(v) if float(v).is_integer() else v


def pulisci(nome: str) -> str:
    """«Show left icon#5615:73» → «Show left icon»"""
    return str(nome).split("#")[0].strip()


def camel(nome: str) -> str:
    parti = [p for p in re.split(r"[\s_-]+", pulisci(nome)) if p]
    if not parti:
        return ""
    return parti[0].lower() + "".join(p[0].upper() + p[1:] for p in parti[1:])


def combinazione(nome: str) -> dict:
    """«Size=md, State=Default» → {Size: md, State: Default}"""
    fuori = {}
    for pezzo in str(nome).split(","):
        parti = [x.strip() for x in pezzo.split("=")]
        if len(parti) == 2:
            fuori[parti[0]] = parti[1]
    return fuori


def ordina(o):
    """Forma canonica: chiavi in ordine, ricorsivo. Gli array mantengono l'ordine."""
    if isinstance(o, list):
        return [ordina(x) for x in o]
    if isinstance(o, dict):
        return {k: ordina(o[k]) for k in sorted(o)}
    return o


def superfici(nodo, nome_var):
    """I colori legati a variabili: sfondo, bordo, testo, icona."""
    fuori = {}
    bv = nodo.get("boundVariables") or {}
    fills = nodo.get("fills") or []
    strokes = nodo.get("strokes") or []
    riempimento = ((fills[0] if fills else {}).get("boundVariables") or {}).get("color", {}).get("id")
    bordo = ((strokes[0] if strokes else {}).get("boundVariables") or {}).get("color", {}).get("id")
    if riempimento:
        fuori["bg"] = nome_var(riempimento)
    if bordo:
        fuori["border"] = nome_var(bordo)

    def scendi(n):
        for c in n.get("children") or []:
            cf = ((c.get("fills") or [{}])[0].get("boundVariables") or {}).get("color", {}).get("id")
            if cf:
                if c.get("type") == "TEXT" and "text" not in fuori:
                    fuori["text"] = nome_var(cf)
                elif "icon" not in fuori:
                    fuori["icon"] = nome_var(cf)
            scendi(c)

    scendi(nodo)
    if bv.get("strokeWeight"):
        fuori["_borderWidth"] = nome_var(bv["strokeWeight"]["id"])
    return fuori


def geometria(nodo, nome_var):
    """Spaziature e raggi legati a variabili, piu' l'altezza misurata."""
    bv = nodo.get("boundVariables") or {}
    g = {}
    mappa = {"paddingTop": "paddingVertical", "paddingLeft": "paddingHorizontal",
             "itemSpacing": "gap", "strokeWeight": "borderWidth"}
    for k, etichetta in mappa.items():
        if bv.get(k):
            g[etichetta] = nome_var(bv[k]["id"])
    raggio = bv.get("topLeftRadius") or bv.get("bottomLeftRadius")
    if raggio:
        g["borderRadius"] = nome_var(raggio["id"])
    riquadro = nodo.get("absoluteBoundingBox") or {}
    g["_measured"] = {"height": arrotonda(riquadro.get("height")), "layout": nodo.get("layoutMode")}
    return g


def main(argv):
    slug = next((a for a in argv if not a.startswith("--")), "button")
    resto = [a for a in argv if not a.startswith("--")]
    node_id = resto[1] if len(resto) > 1 else "5473-10855"
    cfg = json.loads((ROOT / "config" / "figma-files.json").read_text(encoding="utf-8"))
    ds = resto[2] if len(resto) > 2 else cfg["default"]
    if ds not in cfg["files"]:
        sys.exit(f'\n  ✖ ds sconosciuto: "{ds}" — attesi: {", ".join(cfg["files"])}\n')
    chiave_file = cfg["files"][ds]["fileKey"]

    f = Figma()

    # variabileId → nome, per tradurre i boundVariables
    vmeta = f.get(f"files/{chiave_file}/variables/local")["meta"]
    variabili = vmeta.get("variables") or {}
    nome_var = lambda i: (variabili.get(i) or {}).get("name")  # noqa: E731

    nodi = f.get(f"files/{chiave_file}/nodes?ids={node_id}")
    doc = list(nodi["nodes"].values())[0]["document"]
    verifica_nodo(slug, doc.get("name"), ds=ds, node_id=node_id, forza="--forza" in argv)
    varianti = doc.get("children") or []

    # ── 1. Assi: dalla property table, non dalla prosa ──────────────────────
    defs = doc.get("componentPropertyDefinitions") or {}
    assi, booleani, testi, scambi = [], [], [], []
    for nome_grezzo, d in defs.items():
        voce = {"figmaProperty": pulisci(nome_grezzo), "name": camel(nome_grezzo)}
        if d.get("type") == "VARIANT":
            assi.append({**voce, "values": sorted(d.get("variantOptions") or [])})
        elif d.get("type") == "BOOLEAN":
            booleani.append({**voce, "default": d.get("defaultValue")})
        elif d.get("type") == "TEXT":
            testi.append({**voce, "default": d.get("defaultValue")})
        else:
            scambi.append({**voce, "type": d.get("type")})

    # ── 2. Il default: la PRIMA variante del set e' il riferimento di Figma ──
    prima = combinazione(varianti[0]["name"]) if varianti else {}

    trova = lambda r: next((a for a in assi if re.search(r, a["name"], re.I)), None)  # noqa: E731
    asse_size = trova("size")
    asse_aspetto = trova("appear")
    asse_gerarchia = trova("hierarch")
    asse_stato = trova("state")

    # ── 3. Sizing e token: dai boundVariables dei nodi variante ─────────────
    sizing, token = {}, {}
    da_size = da_token = 0
    for v in varianti:
        combo = combinazione(v.get("name", ""))
        sz = combo.get(asse_size["figmaProperty"]) if asse_size else None
        if sz and sz not in sizing:
            sizing[sz] = geometria(v, nome_var)
            da_size += 1

        if asse_aspetto and asse_gerarchia and asse_stato:
            a = combo.get(asse_aspetto["figmaProperty"])
            g = combo.get(asse_gerarchia["figmaProperty"])
            st = combo.get(asse_stato["figmaProperty"])
            if a is None or g is None or not st:
                continue
            # solo dalla size di riferimento, per non moltiplicare
            if asse_size and combo.get(asse_size["figmaProperty"]) != prima.get(asse_size["figmaProperty"]):
                continue
            token.setdefault(f"{a}/{g}", {})[st] = superfici(v, nome_var)
            da_token += 1

    n_size = len(asse_size["values"]) if asse_size else 0
    combinazioni_attese = (
        (len(asse_aspetto["values"]) if asse_aspetto else 0)
        * (len(asse_gerarchia["values"]) if asse_gerarchia else 0)
        * (len(asse_stato["values"]) if asse_stato else 0)
    )

    contratto = {
        "id": f"ds.{slug}",
        "name": doc.get("name"),
        "source": {"ds": ds, "fileKey": chiave_file, "fileName": cfg["files"][ds]["name"],
                   "nodeId": node_id, "nodeType": doc.get("type")},
        "props": [{
            "name": a["name"],
            "figmaProperty": a["figmaProperty"],
            "values": a["values"],
            "default": prima.get(a["figmaProperty"]),
            "_source": {"values": "figma", "order": "figma-alphabetical", "default": "figma-first-variant"},
        } for a in assi],
        "booleans": booleani, "texts": testi, "swaps": scambi,
        "sizing": sizing, "tokens": token,
        "_coverage": {
            "variants": len(varianti),
            "sizingMeasuredOn": f"{da_size}/{n_size} size",
            "tokensMeasuredOn": f"{da_token} combinazioni su {combinazioni_attese}",
            "note": "I token sono misurati sulla sola size di riferimento: le superfici non variano con la size.",
        },
    }

    cartella = USCITA / slug
    cartella.mkdir(parents=True, exist_ok=True)
    scrivi_json(cartella / "contract.json", ordina(contratto))

    L = "─" * 70
    print(f"\n{L}\n  CONTRATTO da Figma — {doc.get('name')}\n{L}")
    print(f"\n▸ NODO      {doc.get('type')} · {len(varianti)} varianti")
    print("\n▸ ASSI      (dalla property table, non dalla prosa)")
    for a in assi:
        print(f"   {a['name']:<12} figma=\"{a['figmaProperty']}\"  default={prima.get(a['figmaProperty']) or '—'}")
        print(f"      {' | '.join(a['values'])}")
    if booleani:
        print("\n▸ BOOLEAN   " + ", ".join(f"{b['name']}={b['default']}" for b in booleani))
    if testi:
        print("▸ TEXT      " + ", ".join(f"{t['name']}=\"{t['default']}\"" for t in testi))
    if scambi:
        print("▸ SWAP      " + ", ".join(s["name"] for s in scambi))
    print(f"\n▸ SIZING    {', '.join(sizing)}")
    for k, g in sizing.items():
        legami = " · ".join(f"{x}={t}" for x, t in g.items() if not x.startswith("_")) or "(nessun binding)"
        print(f"   {k}: {legami}  [h={g['_measured']['height']}]")
    print(f"\n▸ TOKEN     {len(token)} combinazioni appearance/hierarchy")
    print(f"\n▸ COPERTURA {json.dumps(contratto['_coverage'], ensure_ascii=False)}")
    print(f"\n✅ scritto data/contracts/{slug}/contract.json\n{L}\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))
